function.tf

```diff
@@ -0,0 +1,78 @@
+locals {
+  enable_function = var.enable_function || var.enable_extensions
+}
+
+resource "google_service_account" "cloud_functions" {
+  account_id  = "${var.name}-func"
+  description = "A service account for the Observe Cloud Functions"
+}
+
+resource "google_project_iam_member" "cloud_functions" {
+  for_each = toset(var.function_roles)
+
+  project = data.google_project.this.project_id
+  role    = each.key
+  member  = "serviceAccount:${google_service_account.cloud_functions.email}"
+}
+
+resource "google_pubsub_topic_iam_member" "function_pubsub" {
+  topic  = google_pubsub_topic.this.name
+  role   = "roles/pubsub.publisher"
+  member = local.writer_identity
+}
+
+resource "google_cloudfunctions_function" "this" {
+  count = local.enable_function ? 1 : 0
+
+  name                  = var.name
+  description           = "A cloud function that polls data from the GCP REST API"
+  service_account_email = google_service_account.cloud_functions.email
+
+  runtime = "python310"
+  environment_variables = {
+    "PARENT"   = var.resource
+    "TOPIC_ID" = google_pubsub_topic.this.id
+  }
+
+  available_memory_mb = 512
```

Comment:
This should probably be configurable? If you have a lot of objects to list you may need more memory.